memory/memory/memory.py

```diff
@@ -0,0 +1,120 @@
+import aioredis
+import asyncio
+import base64
+import concurrent
+import json
+import logging
+import os
+import uvloop
+from aiohttp import web
+import kubernetes_asyncio as kube
+
+from hailtop.config import get_deploy_config
+from hailtop.google_storage import GCS
+from hailtop.hail_logging import AccessLogger
+from hailtop.tls import get_in_cluster_server_ssl_context
+from hailtop.utils import AsyncWorkerPool, retry_transient_errors
+from gear import setup_aiohttp_session, rest_authenticated_users_only
+
+uvloop.install()
+
+BATCH_PODS_NAMESPACE = os.environ['HAIL_BATCH_PODS_NAMESPACE']
+log = logging.getLogger('batch')
+routes = web.RouteTableDef()
+
+socket = '/redis/redis.sock'
+
+@routes.get('/healthcheck')
+async def healthcheck(request):  # pylint: disable=unused-argument
+    return web.Response()
+
+
+@routes.get('/api/v1alpha/objects')
+@rest_authenticated_users_only
+async def get_object(request, userdata):
+    filename = request.query.get('q')
+    etag = request.query.get('etag')
+    userinfo = await get_or_add_user(request.app, userdata)
+    username = userdata['username']
+    log.info(f'memory: request for object {filename} from user {username}')
+    result = await get_file_or_none(request.app, username, userinfo, filename, etag)
+    if result is None:
+        raise web.HTTPNotFound()
+    etag, body = result
+    return web.Response(headers={'ETag': etag}, body=body)
+
+
+async def get_or_add_user(app, userdata):
+    users = app['users']
+    username = userdata['username']
+    if username not in users:
+        k8s_client = app['k8s_client']
+        gsa_key_secret = await retry_transient_errors(
+            k8s_client.read_namespaced_secret,
+            userdata['gsa_key_secret_name'],
+            BATCH_PODS_NAMESPACE,
+            _request_timeout=5.0)
+        gsa_key = base64.b64decode(gsa_key_secret.data['key.json']).decode()
+        users[username] = {'fs': GCS(blocking_pool=app['thread_pool'], key=json.loads(gsa_key))}
+    return users[username]
+
+
+def make_redis_key(username, filepath):
+    return f'{ username }_{ filepath }'
+
+
+async def get_file_or_none(app, username, userinfo, filepath, etag):
+    filekey = make_redis_key(username, filepath)
+    fs = userinfo['fs']
+
+    cached_etag = await app['redis_pool'].execute('HGET', filekey, 'etag', encoding='ascii')
```

Comment:
You could eliminate a call to redis if you put the etag in the redis key.  Your solution has the advantage it removes the old version from the cache when caching the new one, however, I'm not sure how big a deal that is since the old version will just expire out of the cache anyway.

However, this solution does have a potential race condition:

 - read the cached etag for filekey
 - asynchronously, cache a new version of filekey
 - check the etag and cached etag: they match!
 - read the new body which is now wrong